### ISSUE TITLE
Error using GDAL Virtual File Systems

### ISSUE THREAD

**Issue by wenjieji86** [OSGeo/gdal#2227]
Hi,

I am trying to open and read tif images I stored on Google Cloud storage using the virtual file systems.  My GDAL version is 2.2.3 (somehow failed to upgrade on Google Colab).
Here's what I have tried:
`!gdalinfo --config GS_ACCESS_KEY_ID 'myID' --config GS_SECRET_ACCESS_KEY 'myKEY' /vsigs/bucket_name/subfolder/image.tif`
and I would get:
`ERROR 11: HTTP response code: 403 - Failed writing header
gdalinfo failed - unable to open '/vsigs/bucket_name/subfolder/image.tif'.`
Then I tried:
`!gdalinfo /vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif`
which I would get 1 random error out of 2:
`ERROR 1: Line 1255: </head> doesn't have matching <head>.
gdalinfo failed - unable to open '/vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif'`
or:
`ERROR 3: Cannot read 74247 bytes
gdalinfo failed - unable to open '/vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif'`

I have also tried:
`!gdalinfo --config CPL_VSIL_CURL_USE_HEAD FALSE /vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif`
and I would get:
`ERROR 4: '/vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif' not recognized as a supported file format.
gdalinfo failed - unable to open '/vsicurl/https://storage.cloud.google.com/bucket_name/subfolder/image.tif'.`

I am new to gdal and python, I need help. Thanks.

**Comment by rouault** [comment]
- Is there some non-ASCII characters in "/bucket_name/subfolder/image.tif" ?
- I'm also surprised by seeing a mention of "https://storage.cloud.google.com/". It should be "https://storage.googleapis.com/". Do you override CPL_GS_ENDPOINT ?

**Comment by wenjieji86** [comment]
@rouault no, just letters and '_' in folder and file names.
storage.googleapis.com did not work either. This is interesting because the public url generated by google cloud storage api starts with 'storage.googleapis.com', but that doensn't work for google tensorflow keras api (tf.keras.utils.get_file) it only works if I use 'storage.cloud.google.com'

'CPL_GS_ENDPOINT' is not mentioned in gdal documentation. What is it for? I tried set the variable to both 'storage.googleapis.com' and 'storage.cloud.google.com' and did not work...

**Comment by rouault** [comment]
Do your settings work with google cloud client "gsutil ls gs://your_bucket" ?

**Comment by wenjieji86** [comment]
Yes, it works and I was able to download all files using
google.cloud.storage.Blob(file.name,gcs_bucket).download_to_file(file_obj)

**Comment by wenjieji86** [comment]
@rouault 
Yes, it works and I was able to download all files using
google.cloud.storage.Blob(file.name,gcs_bucket).download_to_file(file_obj)

**Comment by rouault** [comment]
Define CPL_CURL_VERBOSE=YES as environment variable, re-run and post output

**Comment by wenjieji86** [comment]
seems like authentication problem...?
Here's the output:

* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
*   Trying 172.217.203.128...
* TCP_NODELAY set
* Connected to storage.googleapis.com (172.217.203.128) port 443 (#0)
* found 133 certificates in /etc/ssl/certs/ca-certificates.crt
* found 399 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_ECDSA_CHACHA20_POLY1305
* 	 server certificate verification OK
* 	 server certificate status verification SKIPPED
* 	 common name: *.storage.googleapis.com (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: EC/ECDSA
* 	 certificate version: #3
* 	 subject: C=US,ST=California,L=Mountain View,O=Google LLC,CN=*.storage.googleapis.com
* 	 start date: Tue, 21 Jan 2020 08:14:37 GMT
* 	 expire date: Tue, 14 Apr 2020 08:14:37 GMT
* 	 issuer: C=US,O=Google Trust Services,CN=GTS CA 1O1
* 	 compression: NULL
* ALPN, server accepted to use http/1.1
> GET /bucket_name?delimiter=/&prefix=subfolder/ HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:08:50 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:3hDeMFybClIosrkFDKHA8pc4E08=

< HTTP/1.1 403 Forbidden
< X-GUploader-UploadID: AEnB2UrMUWGfWGOfKhTUf1dqsXaaKsdOX7fzsroDIwGknpbGk8Hhb-8J_1Bc7omwUD7eK0R08s6qcmqpwG7vyFQzzjebrO6J0Q
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 310
< Date: Mon, 10 Feb 2020 18:08:51 GMT
< Expires: Mon, 10 Feb 2020 18:08:51 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
< 
* Connection #0 to host storage.googleapis.com left intact
* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
* Found bundle for host storage.googleapis.com: 0x1492f60 [can pipeline]
* Re-using existing connection! (#0) with host storage.googleapis.com
* Connected to storage.googleapis.com (172.217.203.128) port 443 (#0)
> HEAD /bucket_name/subfolder/image.tif HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:08:51 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:iMTWjMV4Ul/R29a320UL+4Hd7ao=

< HTTP/1.1 403 Forbidden
< X-GUploader-UploadID: AEnB2UqlMXYNSeK06YO5c0TiqsKUl6bTHQR6DvHoRSWF-U2qhIitVWEHCd83cxPgLmPyZGEeWpKEdE93mT_LxAghfsrXnXRop1eVPOcTKpndTSHBP76Rn0g
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 331
< Date: Mon, 10 Feb 2020 18:08:51 GMT
< Expires: Mon, 10 Feb 2020 18:08:51 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
* Failed writing header
* stopped the pause stream!
* Closing connection 0
* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
* Hostname storage.googleapis.com was found in DNS cache
*   Trying 172.217.203.128...
* TCP_NODELAY set
* Connected to storage.googleapis.com (172.217.203.128) port 443 (#1)
* found 133 certificates in /etc/ssl/certs/ca-certificates.crt
* found 399 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL re-using session ID
* SSL connection using TLS1.2 / ECDHE_ECDSA_CHACHA20_POLY1305
* 	 server certificate verification OK
* 	 server certificate status verification SKIPPED
* 	 common name: *.storage.googleapis.com (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: EC/ECDSA
* 	 certificate version: #3
* 	 subject: C=US,ST=California,L=Mountain View,O=Google LLC,CN=*.storage.googleapis.com
* 	 start date: Tue, 21 Jan 2020 08:14:37 GMT
* 	 expire date: Tue, 14 Apr 2020 08:14:37 GMT
* 	 issuer: C=US,O=Google Trust Services,CN=GTS CA 1O1
* 	 compression: NULL
* ALPN, server accepted to use http/1.1
> HEAD /bucket_name/subfolder/image.tif.xml HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:08:51 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:/M1Yjl7N0f/NarmPs2x9B1MwmDo=

< HTTP/1.1 403 Forbidden
< X-GUploader-UploadID: AEnB2Up5FBzBvP4Qh_OadwtEizFbkqrVsoAbyg44vQxWQ1c87YLoSBXexnpn6ehDkazYeSDdtBfGvTDU5sKfk3rboN7IKjOhiw
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 335
< Date: Mon, 10 Feb 2020 18:08:51 GMT
< Expires: Mon, 10 Feb 2020 18:08:51 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
* Failed writing header
* stopped the pause stream!
* Closing connection 1
ERROR 11: HTTP response code: 403 - Failed writing header
gdalinfo failed - unable to open '/vsigs/bucket_name/subfolder/image.tif'.

**Comment by rouault** [comment]
> seems like authentication problem...?

yes. Check your GS_ACCESS_KEY_ID and GS_SECRET_ACCESS_KEY values. Maybe you've switched them ?

**Comment by wenjieji86** [comment]
@rouault  oops, forgot to change the bucket name and ran the command....

Here's what I got after re-run:

* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
*   Trying 173.194.216.128...
* TCP_NODELAY set
* Connected to storage.googleapis.com (173.194.216.128) port 443 (#0)
* found 133 certificates in /etc/ssl/certs/ca-certificates.crt
* found 399 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_ECDSA_CHACHA20_POLY1305
* 	 server certificate verification OK
* 	 server certificate status verification SKIPPED
* 	 common name: *.storage.googleapis.com (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: EC/ECDSA
* 	 certificate version: #3
* 	 subject: C=US,ST=California,L=Mountain View,O=Google LLC,CN=*.storage.googleapis.com
* 	 start date: Tue, 21 Jan 2020 08:14:37 GMT
* 	 expire date: Tue, 14 Apr 2020 08:14:37 GMT
* 	 issuer: C=US,O=Google Trust Services,CN=GTS CA 1O1
* 	 compression: NULL
* ALPN, server accepted to use http/1.1
> GET /test_bl%2C_data?delimiter=/&prefix=training/ HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:12:52 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:7JumXH2GlY8c9I3QUKC82G3gH8c=

< HTTP/1.1 400 Bad Request
< X-GUploader-UploadID: AEnB2UqpCtkLyZlLaP7io1H211Z8McvI-LMYSIHT6Tk5cOEEFEww271vAyM8PyurreR2h8n2N3jvzsDfYxnSaWxeTN8OjMbRsw
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 191
< Date: Mon, 10 Feb 2020 18:12:52 GMT
< Expires: Mon, 10 Feb 2020 18:12:52 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
< 
* Connection #0 to host storage.googleapis.com left intact
* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
* Found bundle for host storage.googleapis.com: 0x12d8f90 [can pipeline]
* Re-using existing connection! (#0) with host storage.googleapis.com
* Connected to storage.googleapis.com (173.194.216.128) port 443 (#0)
> HEAD /test_bl%2C_data/training/image.tif HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:12:52 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:NoAQ1JquKfJlfUmI1i5CjBhyRxQ=

< HTTP/1.1 400 Bad Request
< X-GUploader-UploadID: AEnB2Up4o8xHI2BUOpexJMFCV96F_DsbMZKGQ7xs4idWTrjP6jrx3A3BI0f473I3dvK9T-ybESo0B6bZeASt7blKtUxajKywnlYnosEzxf92P_zEZSOBpRQ
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 191
< Date: Mon, 10 Feb 2020 18:12:52 GMT
< Expires: Mon, 10 Feb 2020 18:12:52 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
* Failed writing header
* stopped the pause stream!
* Closing connection 0
* Couldn't find host storage.googleapis.com in the .netrc file; using defaults
* Hostname storage.googleapis.com was found in DNS cache
*   Trying 173.194.216.128...
* TCP_NODELAY set
* Connected to storage.googleapis.com (173.194.216.128) port 443 (#1)
* found 133 certificates in /etc/ssl/certs/ca-certificates.crt
* found 399 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL re-using session ID
* SSL connection using TLS1.2 / ECDHE_ECDSA_CHACHA20_POLY1305
* 	 server certificate verification OK
* 	 server certificate status verification SKIPPED
* 	 common name: *.storage.googleapis.com (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: EC/ECDSA
* 	 certificate version: #3
* 	 subject: C=US,ST=California,L=Mountain View,O=Google LLC,CN=*.storage.googleapis.com
* 	 start date: Tue, 21 Jan 2020 08:14:37 GMT
* 	 expire date: Tue, 14 Apr 2020 08:14:37 GMT
* 	 issuer: C=US,O=Google Trust Services,CN=GTS CA 1O1
* 	 compression: NULL
* ALPN, server accepted to use http/1.1
> HEAD /test_bl%2C_data/training/image.tif.xml HTTP/1.1
Host: storage.googleapis.com
Accept: */*
Date: Mon, 10 Feb 2020 18:12:52 GMT
Authorization: GOOG1 921140803720-8im5gq3tc0cm5enu6t0u8l1beq83qu7l.apps.googleusercontent.com:Wfcb7jJ9auBR+UyMtEZRJp1GjDU=

< HTTP/1.1 400 Bad Request
< X-GUploader-UploadID: AEnB2Up8bShmv0iq6N9LpOU73aGp63pyoFjGu65yizGgXdLXjTqsGrUmQOYlXenK7jsMZbrsicJD5wnJDm2nR6kfTMuoyvbUZ-hoEpCie3IArCSucwVs0b4
< Content-Type: application/xml; charset=UTF-8
< Content-Length: 191
< Date: Mon, 10 Feb 2020 18:12:52 GMT
< Expires: Mon, 10 Feb 2020 18:12:52 GMT
< Cache-Control: private, max-age=0
< Server: UploadServer
* Failed writing header
* stopped the pause stream!
* Closing connection 1
ERROR 11: HTTP response code: 400 - Failed writing header
gdalinfo failed - unable to open '/vsigs/test_bl,_data/training/image.tif'.

**Comment by rouault** [comment]
Ah so your bucket name is "test_bl,_data" with a comma ?

**Comment by wenjieji86** [comment]
@rouault 
My bad. I tried again with correct bucket name and still got the 403 error...
I believe the GS_ACCESS_KEY_ID and GS_SECRET_ACCESS_KEY values are correct - I got them from Google Cloud console credentials page.

Does it matter than I am running my code in colab and I have authenticated using the following code (to use the gcs client apis):
`from google.colab import auth
auth.authenticate_user()`

**Comment by rouault** [comment]
I've no knowledge of Colab, but I fail to see how it would interact with GDAL. GDAL doesn't its own thing. I'm a bit at loss if it is just a user issue on your side or some real bug. You're not running the latest GDAL version either, so maybe there would be some later fixed bug, but digging a bit in history doesn't show anything particularly relevant.
Perhaps try outside of Colab in a normal Linux or Windows environment to see if you can reproduce

**Comment by wenjieji86** [comment]
@rouault 
Now I am pretty sure it's an authentication issue.
I tired on my local PC (python 3.6 + GDAL 3.0) and ran into the same issue...
Looking into GCS authentical...
Thanks!

**Comment by wenjieji86** [comment]
@rouault 
OK, I figured it out...
I should have gotten the id and the secret for a service account of my project rather than creating an OAuth Client ID and secret...

Thank you for your help!
